pkg/geo/geomfn/subdivide.go

```diff
@@ -0,0 +1,364 @@
+		}
+		return nil
+	}
+
+	splitHorizontally := width > height
+	splitPoint, err := calculateSplitPoint(gt, nVertices, splitHorizontally, clip)
+	if err != nil {
+		return err
+	}
+	subBox1 := *clip
+	subBox2 := *clip
+	if splitHorizontally {
+		subBox1.HiX = splitPoint
+		subBox2.LoX = splitPoint
+	} else {
+		subBox1.HiY = splitPoint
+		subBox2.LoY = splitPoint
+	}
+
+	clipped1, err := clipGeomTByBoundingBox(gt, &subBox1)
+	if err != nil {
+		return errors.Newf("error clipping geom.T: %v", err)
+	}
+	clipped2, err := clipGeomTByBoundingBox(gt, &subBox2)
+	if err != nil {
+		return errors.Newf("error clipping geom.T: %v", err)
+	}
+	depth++
+	err = subdivideRecursive(clipped1, maxVertices, depth, col, dim)
+	if err != nil {
+		return err
+	}
+	err = subdivideRecursive(clipped2, maxVertices, depth, col, dim)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// isGeomTCollection returns whether the given geom.T is of a collection type (wrapper for a geometry function).
+func isGeomTCollection(gt geom.T) (bool, error) {
+	g, err := geo.MakeGeometryFromGeomT(gt)
+	if err != nil {
+		return false, err
+	}
+	return IsCollection(g)
+}
+
+// forEachGeomTInCollection calls the provided function for each geometry in the passed in collection.
+// Ignores MultiPoint input.
+func forEachGeomTInCollection(gt geom.T, fn func(geom.T) error) error {
+	switch gt := gt.(type) {
+	case *geom.GeometryCollection:
+		for i := 0; i < gt.NumGeoms(); i++ {
+			err := fn(gt.Geom(i))
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	case *geom.MultiPolygon:
+		for i := 0; i < gt.NumPolygons(); i++ {
+			err := fn(gt.Polygon(i))
+			if err != nil {
+				return err
+			}
+		}
+	case *geom.MultiLineString:
+		for i := 0; i < gt.NumLineStrings(); i++ {
+			err := fn(gt.LineString(i))
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func calculateSplitPoint(
+	gt geom.T, nVertices int, splitHorizontally bool, clip *geo.CartesianBoundingBox,
+) (float64, error) {
+	var pivot float64
+	switch gt := gt.(type) {
+	case *geom.Polygon:
+		var err error
+		pivot, err = findPolygonsMostCentralPointValue(gt, nVertices, splitHorizontally, clip)
+		if err != nil {
+			return 0, errors.Newf("error finding most central point for polygon: %v", err)
+		}
+		if splitHorizontally {
+			// Ignore point on the boundaries
+			if clip.LoX == pivot || clip.HiX == pivot {
+				pivot = (clip.LoX + clip.HiX) / 2
+			}
+		} else {
+			// Ignore point on the boundaries
+			if clip.LoY == pivot || clip.HiY == pivot {
+				pivot = (clip.LoY + clip.HiY) / 2
+			}
+		}
+	default:
+		if splitHorizontally {
+			pivot = (clip.LoX + clip.HiX) / 2
+		} else {
+			pivot = (clip.LoY + clip.HiY) / 2
+		}
+	}
+	return pivot, nil
+}
+
+// findPolygonsMostCentralPointValue finds the value of the most central point for a geom.Polygon.
+// Can be X or Y value, depending on the splitHorizontally argument.
+func findPolygonsMostCentralPointValue(
```

Comment:
nit: `findMostCentralPointForPolygon` (main grit is the plural)

Reply:
can you also comment on why we can't use centroid?

Reply:
Fixed.
Using the central point of an input geometry gives us "simpler" geometry shapes in the results (comparing to the centroid). Imagine a `Polygon` created by joining a triangle and a square twice it's size, our goal being to keep only geometries with less than 5 vertices. Splitting this input exactly in half would give us a geometry of a similar shape (still 5 vertices) and a small rectangle or two polygons with 5 vertices. By splitting using the most central point, we can get a triangle and a square, that already satisfy the exit condition.